# Data Science I Topic 2

## DataFrames

### From a dictionary

<u>In T0-TupleListDictionary, you've seen the following dictionary. Run the cell.</u>

In [1]:
movie_ratings = {'Toy Story':{'rating':4.0, 'genre':'Animation'},
                          'Jumanji':{'rating':4.0, 'genre':'Adventure'},
                          'Grumpier Old Men':{'rating':4.0, 'genre':'Comedy'},
                          'Waiting to Exhale':{'rating':4.0, 'genre':'Comedy'}
                         }

<u>Let's now create a simple DataFrame from this dictionary using [Pandas](https://pandas.pydata.org/) package. Follow the instructions below.</u>

In [2]:
# import pandas package with the alias pd
import pandas as pd

In [3]:
ratings_df = pd.DataFrame(movie_ratings)

# display ratings_df#
print(ratings_df)

        Toy Story    Jumanji Grumpier Old Men Waiting to Exhale
rating        4.0        4.0              4.0               4.0
genre   Animation  Adventure           Comedy            Comedy


### Transposing a DataFrame

<u>With this simple DataFrame, let's say we want the rating and genre to be the **columns/variables**.</u>

In [4]:
# How to use transpose() on a DataFrame? overwrite ratings_df with its transpose
ratings_df = ratings_df.transpose()

# print ratings_df
print(ratings_df)

                  rating      genre
Toy Story            4.0  Animation
Jumanji              4.0  Adventure
Grumpier Old Men     4.0     Comedy
Waiting to Exhale    4.0     Comedy


<u>The title of the movies became the **index** of the DataFrame.</u>

### Using `reset_index` and  `set_index`

<u>What if we don't want the title to be the indices, instead contained in a column?<br>The current index has no name, and we want the contents to be under the column `title`. So we'll first rename it, before using `reset_index`.</u>

In [5]:
ratings_df = ratings_df.rename_axis('title').reset_index()

# print ratings_df
print(ratings_df)

               title rating      genre
0          Toy Story    4.0  Animation
1            Jumanji    4.0  Adventure
2   Grumpier Old Men    4.0     Comedy
3  Waiting to Exhale    4.0     Comedy


<u>You can simply use `.set_index('title')` on ratings_df to reverse this.<br>We will see later, when using a bigger DataFrame, how `index` and `columns` can be useful.</u>

***

## Reading csv files

<u>We will use the [MovieLens Latest Datasets](https://grouplens.org/datasets/movielens/latest/). Download ml-latest-small.zip and extract the zip file. There are four csv files (`links.csv`, `movies.csv`, `ratings.csv`, and `tags.csv`). Take some time to read the README file.<br>We'll first look at `ratings.csv`</u>

In [6]:
# ONLY IF YOU'RE USING GOOGLE COLAB AND WANT TO MOUNT YOUR DRIVE TO ACCESS YOUR FILES,
# uncomment these two lines:
# from google.colab import drive
# drive.mount('/content/drive')

# Once you're done:
#drive.flush_and_unmount()

# ALTERNATIVE: Upload the file to the session storage (will be deleted once your session ends). 
# Click on the folder icon, then the icon with the up arrow symbol.

In [7]:
# Change the path if necessary
ratings_df = pd.read_csv('ml-latest-small/ratings.csv')

<u>Read the rest of the csv files.</u>

In [8]:
movies_df = pd.read_csv('ml-latest-small/movies.csv')
links_df = pd.read_csv('ml-latest-small/links.csv')
tags_df = pd.read_csv('ml-latest-small/tags.csv')

***

## Exploring DataFrames

<u>Run the cells below, follow the instructions and inspect the outputs.</u>

In [9]:
ratings_df.head(3) # without number specification, the default of 5 entries will be printed

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224

<u>We want the timestamp column to be more readable and easier to deal with. Let's change it to DateTime format.</u>

In [10]:
ratings_df["time"] = pd.to_datetime(ratings_df.timestamp, unit='s')
ratings_df = ratings_df.drop(columns="timestamp")

In [11]:
# .tail() will print the last few entries.
# show the last 3 entries of ratings_df
ratings_df.tail(3)

userId  movieId  rating                time
100833     610   168250     5.0 2017-05-08 19:50:47
100834     610   168252     5.0 2017-05-03 21:19:12
100835     610   170875     3.0 2017-05-03 21:20:15

In [12]:
# show the first few entries of movies_df
movies_df.head(3)

movieId                    title  \
0        1         Toy Story (1995)   
1        2           Jumanji (1995)   
2        3  Grumpier Old Men (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance

In [13]:
# show the first few entries of links_df
links_df.head(3)

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0

In [14]:
# tags_df also contains a Unix timestamp, "timestamp", column

# add a column "time" that contains the DateTime format of "timestamp"
tags_df["time"] = pd.to_datetime(tags_df.timestamp, unit='s')

# drop the column "timestamp"
tags_df = tags_df.drop(columns="timestamp")

# show the first few entries of tags_df
tags_df.head(3)

userId  movieId              tag                time
0       2    60756            funny 2015-10-24 19:29:54
1       2    60756  Highly quotable 2015-10-24 19:29:56
2       2    60756     will ferrell 2015-10-24 19:29:52

<u>Run the following cells and inspect the outputs. How many entries are there in `ratings_df`? Are there any empty (null) entries?</u>

In [15]:
# Print the size 
ratings_df.shape

(100836, 4)

In [16]:
# Print the information about ratings_df
ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype         
---  ------   --------------   -----         
 0   userId   100836 non-null  int64         
 1   movieId  100836 non-null  int64         
 2   rating   100836 non-null  float64       
 3   time     100836 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(2)
memory usage: 3.1 MB


In [17]:
# Print the description 
ratings_df.describe()

userId        movieId         rating
count  100836.000000  100836.000000  100836.000000
mean      326.127564   19435.295718       3.501557
std       182.618491   35530.987199       1.042529
min         1.000000       1.000000       0.500000
25%       177.000000    1199.000000       3.000000
50%       325.000000    2991.000000       3.500000
75%       477.000000    8122.000000       4.000000
max       610.000000  193609.000000       5.000000

<u>Now the userId and movieId are still numerical. We can use .astype(str) or .astype('category')) or pd.Category(...) to change them. But we'll leave them be for now.</u>

In [18]:
# Print the column names
ratings_df.columns

Index(['userId', 'movieId', 'rating', 'time'], dtype='object')

In [19]:
# Print the indices
ratings_df.index

RangeIndex(start=0, stop=100836, step=1)

In [20]:
ratings_df.tail(2)

userId  movieId  rating                time
100834     610   168252     5.0 2017-05-03 21:19:12
100835     610   170875     3.0 2017-05-03 21:20:15

In [21]:
ratings_df.iloc[100835,:]

userId                     610
movieId                 170875
rating                     3.0
time       2017-05-03 21:20:15
Name: 100835, dtype: object

**Ans**: 100836 entries, there's no null entries.

<u>Questions:</u>
1. 

***

## Subsetting and Sorting

### Subsetting columns

<u>Run the cells below.</u>

In [22]:
only_ratings = ratings_df["rating"]

<u>Q: What is the data type?</u>

In [23]:
only_ratings.dtype

dtype('float64')

In [24]:
# print the data type of only_ratings
print(type(only_ratings))

<class 'pandas.core.series.Series'>


<u>We can subset multiple columns by passing the column names as a list.</u>

In [25]:
# Select columns "movieId" and "rating"
print(ratings_df[["movieId", "rating"]])
print(type(ratings_df))

        movieId  rating
0             1     4.0
1             3     4.0
2             6     4.0
3            47     5.0
4            50     5.0
...         ...     ...
100831   166534     4.0
100832   168248     5.0
100833   168250     5.0
100834   168252     5.0
100835   170875     3.0

[100836 rows x 2 columns]
<class 'pandas.core.frame.DataFrame'>


In [26]:
ratings_df[["movieId", "rating"]].head()

movieId  rating
0        1     4.0
1        3     4.0
2        6     4.0
3       47     5.0
4       50     5.0

### Sorting rows

<u>Let's sort the movie based on their ratings.</u>

In [27]:
ratings_df.sort_values("rating")

userId  movieId  rating                time
3752       22    53519     0.5 2010-03-16 08:12:17
60861     393     5445     0.5 2015-05-01 18:57:16
47025     307     2017     0.5 2007-08-03 20:40:39
22446     153     1198     0.5 2018-05-05 19:24:24
60865     393     5902     0.5 2015-05-01 19:11:49
...       ...      ...     ...                 ...
90260     587       50     5.0 2000-03-15 17:29:26
90261     587       58     5.0 2000-03-15 17:33:12
17061     108     5303     5.0 2003-01-17 21:41:59
90266     587      236     5.0 2000-03-15 16:59:01
81530     516     4429     5.0 2001-07-08 20:48:19

[100836 rows x 4 columns]

<u>How to get the highest rating movies on the top?</u>

In [28]:
# Try it: pass the argument ascending=False
ratings_df.sort_values("rating", ascending=False)

userId  movieId  rating                time
34031     232     3147     5.0 2008-08-08 03:08:23
50717     328     1080     5.0 2017-05-08 02:27:30
18166     115      487     5.0 1999-12-11 22:29:38
18165     115      480     5.0 1999-12-11 23:08:53
76157     479     3308     5.0 2002-12-08 15:49:18
...       ...      ...     ...                 ...
99278     608     4679     0.5 2005-05-31 02:30:42
99277     608     4678     0.5 2005-05-31 02:23:22
95668     600     4744     0.5 2009-03-22 09:33:26
99269     608     4621     0.5 2005-05-31 02:35:16
78499     489      168     0.5 2012-03-26 15:11:53

[100836 rows x 4 columns]

### Sorting rows based on multiple columns

<u>To sort the rows based on multiple columns, pass the column names and the `ascending` argument as lists.</u>

In [29]:
# Try it: with descending rating and ascending time
ratings_df.sort_values(["rating", "time"], ascending=[False, True])

userId  movieId  rating                time
66665     429      150     5.0 1996-03-29 18:36:55
66667     429      161     5.0 1996-03-29 18:36:55
66716     429      588     5.0 1996-03-29 18:36:55
66717     429      590     5.0 1996-03-29 18:36:55
66718     429      592     5.0 1996-03-29 18:36:55
...       ...      ...     ...                 ...
81446     514   141994     0.5 2018-09-08 04:35:03
58089     380   179819     0.5 2018-09-13 21:05:21
27248     184   184641     0.5 2018-09-16 10:46:48
27232     184   175475     0.5 2018-09-16 14:52:50
51251     331     6534     0.5 2018-09-17 04:26:18

[100836 rows x 4 columns]

### Subsetting rows using a Boolean series

<u>Run the cells below.</u>

In [30]:
# This will result in a Boolean series
ratings_df["userId"] == 1 

0          True
1          True
2          True
3          True
4          True
          ...  
100831    False
100832    False
100833    False
100834    False
100835    False
Name: userId, Length: 100836, dtype: bool

In [31]:
# Filtering for rows where userId=1
ratings_df[ratings_df["userId"]==1]

userId  movieId  rating                time
0         1        1     4.0 2000-07-30 18:45:03
1         1        3     4.0 2000-07-30 18:20:47
2         1        6     4.0 2000-07-30 18:37:04
3         1       47     5.0 2000-07-30 19:03:35
4         1       50     5.0 2000-07-30 18:48:51
..      ...      ...     ...                 ...
227       1     3744     4.0 2000-07-30 18:11:34
228       1     3793     5.0 2000-07-30 18:30:55
229       1     3809     4.0 2000-07-30 18:20:20
230       1     4006     4.0 2000-07-30 18:48:23
231       1     5060     5.0 2000-07-30 19:06:42

[232 rows x 4 columns]

<u>Now create a DataFrame called `high_ratings_df` where it only contains the entries with rating=5.0.</u>

In [32]:
# Filtering for rows where rating=5.0
high_ratings_df = ratings_df[ratings_df["rating"]==5.0]

# Print high_ratings_df
print(high_ratings_df)

        userId  movieId  rating                time
3            1       47     5.0 2000-07-30 19:03:35
4            1       50     5.0 2000-07-30 18:48:51
6            1      101     5.0 2000-07-30 18:14:28
8            1      151     5.0 2000-07-30 19:07:21
9            1      157     5.0 2000-07-30 19:08:20
...        ...      ...     ...                 ...
100814     610   158238     5.0 2016-11-19 08:46:59
100829     610   164179     5.0 2017-05-03 21:07:11
100832     610   168248     5.0 2017-05-03 22:21:31
100833     610   168250     5.0 2017-05-08 19:50:47
100834     610   168252     5.0 2017-05-03 21:19:12

[13211 rows x 4 columns]


### Subsetting rows with multiple conditions

<u>Which movieId did user 1 give 5.0 rating to?</u>

In [33]:
# Construct the Boolean series
is_user1 = ratings_df["userId"] == 1
is_rating5 = ratings_df["rating"] == 5.0

# Try it: Use both Boolean series to subset ratings_df. Which logical operator do you need?
ratings_df[is_user1 & is_rating5]

userId  movieId  rating                time
3         1       47     5.0 2000-07-30 19:03:35
4         1       50     5.0 2000-07-30 18:48:51
6         1      101     5.0 2000-07-30 18:14:28
8         1      151     5.0 2000-07-30 19:07:21
9         1      157     5.0 2000-07-30 19:08:20
..      ...      ...     ...                 ...
223       1     3702     5.0 2000-07-30 19:02:17
224       1     3703     5.0 2000-07-30 18:31:49
225       1     3729     5.0 2000-07-30 18:39:23
228       1     3793     5.0 2000-07-30 18:30:55
231       1     5060     5.0 2000-07-30 19:06:42

[124 rows x 4 columns]

numpy arrays and pandas series use the bitwise operators rather than logical as you are comparing every element in the array/series with another

### Subsetting using `.isin()`

<u>Say we want to subset the ratings based on more than two conditions, checking the conditions one by one can get quite tedious. In this case, it's better to use `.isin()`.<br>Run the cell below.</u>

In [34]:
# also results in Boolean series
is_movie123 = ratings_df['movieId'].isin([1, 2, 3])

#subset ratings_df with is_movie123
ratings_df[is_movie123]

userId  movieId  rating                time
0           1        1     4.0 2000-07-30 18:45:03
1           1        3     4.0 2000-07-30 18:20:47
516         5        1     4.0 1996-11-08 06:36:02
560         6        2     4.0 1996-10-17 11:58:42
561         6        3     5.0 1996-10-17 12:11:36
...       ...      ...     ...                 ...
98666     608        1     2.5 2005-05-29 23:11:07
98667     608        2     2.0 2005-05-30 22:06:26
98668     608        3     2.0 2005-05-31 01:53:33
99497     609        1     3.0 1996-11-05 19:10:25
99534     610        1     5.0 2016-11-19 08:08:20

[377 rows x 4 columns]

### Hierarchical indices

<u>We've seen how to use `.set_index()` and `.reset_index()`before, now let's try setting multiple indices.</u>

In [35]:
# create a list of the column names to set as indices
indices = ["userId", "movieId"]

# use .set_index() on ratings_df and pass the indices as the argument
ratings_multindex = ratings_df.set_index(indices)

# print ratings_multindex
ratings_multindex

rating                time
userId movieId                            
1      1           4.0 2000-07-30 18:45:03
       3           4.0 2000-07-30 18:20:47
       6           4.0 2000-07-30 18:37:04
       47          5.0 2000-07-30 19:03:35
       50          5.0 2000-07-30 18:48:51
...                ...                 ...
610    166534      4.0 2017-05-03 21:53:22
       168248      5.0 2017-05-03 22:21:31
       168250      5.0 2017-05-08 19:50:47
       168252      5.0 2017-05-03 21:19:12
       170875      3.0 2017-05-03 21:20:15

[100836 rows x 2 columns]

### Subsetting with indices

<u>Here we'll learn how to use `loc` and `iloc` to subset a DataFrame using its indices.
We'll use `tags_df`.
</u>

#### Outer index level

In [36]:
print(tags_df.head(2))

# Set tag as the index
tags_1idx = tags_df.set_index('tag')
tags_1idx.head()

   userId  movieId              tag                time
0       2    60756            funny 2015-10-24 19:29:54
1       2    60756  Highly quotable 2015-10-24 19:29:56


userId  movieId                time
tag                                                 
funny                 2    60756 2015-10-24 19:29:54
Highly quotable       2    60756 2015-10-24 19:29:56
will ferrell          2    60756 2015-10-24 19:29:52
Boxing story          2    89774 2015-10-24 19:33:27
MMA                   2    89774 2015-10-24 19:33:20

In [37]:
# print out all indices
tags_1idx.index

Index(['funny', 'Highly quotable', 'will ferrell', 'Boxing story', 'MMA',
       'Tom Hardy', 'drugs', 'Leonardo DiCaprio', 'Martin Scorsese',
       'way too long',
       ...
       'music', 'British', 'Romans', '70mm', 'World War II', 'for katie',
       'austere', 'gun fu', 'heroic bloodshed', 'Heroic Bloodshed'],
      dtype='object', name='tag', length=3683)

<u> You've seen the following examples in the lecture</u>

In [38]:
# iloc: by the integer position(s)
tags_1idx.iloc[[2,0]]

userId  movieId                time
tag                                              
will ferrell       2    60756 2015-10-24 19:29:52
funny              2    60756 2015-10-24 19:29:54

In [39]:
# loc: by the label(s)
tags_1idx.loc[['funny', 'will ferrell']].tail()

userId  movieId                time
tag                                              
funny            599     1732 2017-06-26 05:51:31
will ferrell       2    60756 2015-10-24 19:29:52
will ferrell      62    60756 2018-06-13 23:59:39
will ferrell      62   107348 2018-06-14 00:10:02
will ferrell     424    60756 2016-03-13 05:15:29

#### Inner index level

<u>We'll set a second index based on the year of ratings.</u>

In [40]:
tags_df["year"] = tags_df.time.dt.year

tags_2idx = tags_df.set_index(['tag', 'year']).sort_index()
tags_2idx.head(10)

userId  movieId  \
tag                                       year                    
"artsy"                                   2018     567     4552   
06 Oscar Nominated Best Movie - Animation 2006     474    31658   
                                          2006     474    37729   
                                          2006     474    38038   
1900s                                     2006     474      918   
1920s                                     2006     474     1084   
                                          2006     474     3548   
1950s                                     2006     474     1103   
                                          2006     474     2664   
1960s                                     2006     474     1093   

                                                              time  
tag                                       year                      
"artsy"                                   2018 2018-05-02 18:31:18  
06 Oscar Nominated Best Movie - Animation 2006 2006-04-19 13:04:16  
                                          2006 2006-02-04 13:32:18  
                                          2006 2006-02-11 13:23:10  
1900s                                     2006 2006-01-24 21:25:49  
1920s                                     2006 2006-01-24 21:24:04  
                                          2006 2006-01-14 02:22:11  
1950s                                     2006 2006-01-14 02:51:28  
                                          2006 2006-01-23 16:00:17  
1960s                                     2006 2006-01-16 01:42:16

In [41]:
# Try it: Pass a tag and a year to .loc[]
tags_2idx.loc["funny", 2015]

userId  movieId                time
tag   year                                     
funny 2015       2    60756 2015-10-24 19:29:54
      2015     119   101142 2015-07-10 21:17:47
      2015     256   126548 2015-11-14 20:23:12
      2015     537    69122 2015-02-17 02:31:57

<u> Run the cells and inspect the results</u>

In [42]:
# Slicing based on multiple indices
print(tags_2idx.loc[("funny",2015):("funny",2017)])
print("---------------------")
print(tags_2idx.loc[("will ferrell",(2015,2018)),])

            userId  movieId                time
tag   year                                     
funny 2015       2    60756 2015-10-24 19:29:54
      2015     119   101142 2015-07-10 21:17:47
      2015     256   126548 2015-11-14 20:23:12
      2015     537    69122 2015-02-17 02:31:57
      2016     424    60756 2016-03-13 05:15:27
      2017     599      296 2017-06-26 05:53:03
      2017     599     1732 2017-06-26 05:51:31
---------------------
                   userId  movieId                time
tag          year                                     
will ferrell 2015       2    60756 2015-10-24 19:29:52
             2018      62    60756 2018-06-13 23:59:39
             2018      62   107348 2018-06-14 00:10:02


In [43]:
# Inspect the results
print(tags_2idx.loc["Al Pacino":"Alicia Vikander"])
print("---------------------")
print(tags_2idx.loc[[("Al Pacino",2018),("Alicia Vikander",2018)], ["userId", "movieId"]])

                       userId  movieId                time
tag              year                                     
Al Pacino        2016      18      431 2016-05-01 21:39:25
                 2016      18     1221 2016-04-26 19:35:06
                 2016     424     4262 2016-03-13 20:29:25
                 2018      62     2023 2018-05-06 19:58:48
                 2018      62     5388 2018-06-29 22:20:09
Alcatraz         2006     474      733 2006-01-14 02:40:18
Alfred Hitchcock 2009     477      903 2009-05-17 16:18:45
                 2009     477      908 2009-05-17 16:18:17
                 2009     477     1219 2009-05-17 16:19:13
                 2016     424     1219 2016-03-13 20:54:15
Alicia Vikander  2018      62   184471 2018-06-03 11:21:54
---------------------
                      userId  movieId
tag             year                 
Al Pacino       2018      62     2023
                2018      62     5388
Alicia Vikander 2018      62   184471


### Subsetting in both directions

<u>This is especially useful if your DataFrame has many columns (variables).</u>

In [44]:
# Try it: display the columns "userId" and "movieId" 
# for movies with tag "funny" between years 2016-2018
# Pass the columns as a tuple
print( tags_2idx.loc[("funny",2016):("funny",2018), 
                     ("userId", "movieId")] )

            userId  movieId
tag   year                 
funny 2016     424    60756
      2017     599      296
      2017     599     1732
      2018      62     2953
      2018      62     3114
      2018      62    60756
      2018      62    68848
      2018      62    71535
      2018      62    88405
      2018      62    99114
      2018      62   119141
      2018      62   179401
      2018      62   183611
      2018     567   106766
      2018     567   112852
      2018     567   134170
      2018     567   148626
      2018     567   167746


### Slicing by (partial) dates

<u>Say we want to see only the ratings made in 2018 only.</u>

In [45]:
# Try it: get the Boolean series using logical operator/s on ratings_df["time"]
is_year2018 = (ratings_df["time"]>="2018") & (ratings_df["time"]<"2019")

# ratings made in 2018
ratings_df[is_year2018]

userId  movieId  rating                time
1839       18      913     4.0 2018-02-01 22:57:29
1852       18     1193     4.5 2018-02-03 19:27:05
1980       18     5120     4.5 2018-01-15 22:08:35
2012       18     6787     4.0 2018-08-11 20:54:56
2053       18    27878     4.5 2018-05-12 21:18:13
...       ...      ...     ...                 ...
95960     601   170705     5.0 2018-03-18 18:26:36
95961     601   172591     4.5 2018-03-19 13:56:59
95962     601   174055     4.0 2018-03-18 18:28:59
95963     601   176371     4.0 2018-03-18 18:27:03
95964     601   177765     4.5 2018-03-18 18:27:01

[6418 rows x 4 columns]

In [46]:
# Alternative: use .dt.year
# Get the Boolean series
is_year2018 = ratings_df["time"].dt.year==2018

# ratings made in 2018
ratings_df[is_year2018]

userId  movieId  rating                time
1839       18      913     4.0 2018-02-01 22:57:29
1852       18     1193     4.5 2018-02-03 19:27:05
1980       18     5120     4.5 2018-01-15 22:08:35
2012       18     6787     4.0 2018-08-11 20:54:56
2053       18    27878     4.5 2018-05-12 21:18:13
...       ...      ...     ...                 ...
95960     601   170705     5.0 2018-03-18 18:26:36
95961     601   172591     4.5 2018-03-19 13:56:59
95962     601   174055     4.0 2018-03-18 18:28:59
95963     601   176371     4.0 2018-03-18 18:27:03
95964     601   177765     4.5 2018-03-18 18:27:01

[6418 rows x 4 columns]

In [47]:
# Try it: Get the ratings made in years 2000 and 2001.
is_interval = (ratings_df["time"].dt.year==2000) | (ratings_df["time"].dt.year==2001)

ratings_df[is_interval]

userId  movieId  rating                time
0           1        1     4.0 2000-07-30 18:45:03
1           1        3     4.0 2000-07-30 18:20:47
2           1        6     4.0 2000-07-30 18:37:04
3           1       47     5.0 2000-07-30 19:03:35
4           1       50     5.0 2000-07-30 18:48:51
...       ...      ...     ...                 ...
98661     607     4022     4.0 2001-08-15 03:46:13
98662     607     4023     3.0 2001-08-15 03:46:13
98663     607     4054     3.0 2001-08-15 03:46:13
98664     607     4069     3.0 2001-08-15 03:46:43
98665     607     5060     3.0 2000-07-08 17:58:51

[13983 rows x 4 columns]

## Merging and Pivoting

<u>In the lecture we have seen the example of merging and pivoting using `tags_df` and `ratings_df`.</u>

In [48]:
# Print a few lines from tags_df and ratings_df to remind ourselves how they look like.
print(tags_df.head(2))
print(ratings_df.head(2))

   userId  movieId              tag                time  year
0       2    60756            funny 2015-10-24 19:29:54  2015
1       2    60756  Highly quotable 2015-10-24 19:29:56  2015
   userId  movieId  rating                time
0       1        1     4.0 2000-07-30 18:45:03
1       1        3     4.0 2000-07-30 18:20:47


In [49]:
# Merge the two DataFrames on two columns, "userId" and "movieId",
# by first dropping the time stamp on ratings_df
tags_ratings_df = tags_df.merge(ratings_df.drop(columns=["time"]), 
                                on=["userId", "movieId"])

tags_ratings_df

userId  movieId               tag                time  year  rating
0          2    60756             funny 2015-10-24 19:29:54  2015     5.0
1          2    60756   Highly quotable 2015-10-24 19:29:56  2015     5.0
2          2    60756      will ferrell 2015-10-24 19:29:52  2015     5.0
3          2    89774      Boxing story 2015-10-24 19:33:27  2015     5.0
4          2    89774               MMA 2015-10-24 19:33:20  2015     5.0
...      ...      ...               ...                 ...   ...     ...
3471     606     6107      World War II 2007-05-06 17:49:07  2007     4.0
3472     606     7382         for katie 2007-02-11 22:46:59  2007     4.5
3473     610     3265            gun fu 2017-05-03 20:39:44  2017     5.0
3474     610     3265  heroic bloodshed 2017-05-03 20:39:38  2017     5.0
3475     610   168248  Heroic Bloodshed 2017-05-03 20:44:30  2017     5.0

[3476 rows x 6 columns]

<u>Read the description in the cell below and run it.</u>

In [52]:
# Pivot table by year and aggregate the ratings with their average
mean_yearly_rating = tags_ratings_df.pivot_table(index="tag", columns="year", 
                                         values="rating", aggfunc="mean")

# This line of code sorts the pivot table by the least occurrence of NAs, 
# because not each year has data, and replace NAs by "-"
mean_yearly_rating.iloc[mean_yearly_rating.isnull().sum(axis=1).argsort()].fillna("-").head(10)

year              2006 2007 2008      2009  2010 2011 2012 2013 2014  \
tag                                                                    
music         3.785714  4.0    -  4.333333   5.0    -    -    -    -   
psychology    3.916667    -    -  4.666667     -    -    -  5.0    -   
dark comedy          -    -  5.0       5.0   4.5    -    -  5.0    -   
mindfuck             -    -    -      4.75   5.0    -  5.0    -    -   
atmospheric          -    -    -       5.0   5.0  5.0    -    -    -   
suspense             -    -    -       4.5     -    -  4.0    -    -   
surreal              -    -    -       4.5  4.75    -    -    -    -   
twist ending       4.0    -    -      4.25     -    -  4.5    -    -   
black comedy         -    -    -       5.0   5.0    -    -    -    -   
sci-fi               -    -    -         -   4.3    -    -    -    -   

year              2015      2016  2017      2018  
tag                                               
music             4.25       3.0   5.0         -  
psychology         5.0  4.416667   5.0      4.75  
dark comedy          -     4.125   5.0  4.142857  
mindfuck           5.0       4.6   5.0  4.166667  
atmospheric        4.5       4.4   5.0  3.954545  
suspense           4.5      4.75   5.0  3.166667  
surreal       3.833333       3.5   5.0  4.192308  
twist ending         -  4.666667   5.0       5.0  
black comedy         -       4.5   5.0  4.444444  
sci-fi             4.0     2.875  4.75  4.333333

In [53]:
# Try it: Create the table of the median rating per year from 2010 onwards,
# set the tag as the index and years as the columns
# display 10 lines, sort based on the number the least number of NAs like before, and 
# replace NAs by blanks ("")

median_yearly_rating = tags_ratings_df[tags_ratings_df.year>=2010].pivot_table(index="tag", columns="year", 
                                         values="rating", aggfunc="median")
median_yearly_rating.iloc[median_yearly_rating.isnull().sum(axis=1).argsort()].fillna("").head(10)

year         2010 2011 2012 2013 2014  2015  2016  2017  2018
tag                                                          
mindfuck      5.0       5.0             5.0   5.0   5.0   4.0
atmospheric   5.0  5.0                  4.5   4.5   5.0  3.75
psychology                   5.0        5.0   4.5   5.0  4.75
surreal      4.75                       4.0   3.5   5.0   4.0
dark comedy   4.5            5.0             4.25   5.0   4.0
stylized      5.0       5.0                  3.75   5.0   4.5
clever                  4.5             5.0   5.0   5.0   4.0
funny                   4.5            4.25   4.0   5.0   3.5
sci-fi        4.0                       4.0  2.25  4.75   4.5
comedy        3.5                       4.0   4.5   5.0   3.5

In [54]:
# Try it: Create the table of the count of ratings
# We're only interested with the tags that start with 'com', e.g. 'comic', 'computer', etc
# set the year as the index and tags as the columns
# display all lines sorted by year
# replace NAs by "-"
count_yearly_rating = tags_ratings_df[tags_ratings_df.tag.str[0:3]=='com'].pivot_table(columns="tag", index="year", 
                                         values="rating", aggfunc="count")
count_yearly_rating.sort_index().fillna("-")

tag  coma comedy comic book comics complicated computer computer animation  \
year                                                                         
2006  2.0      -          -      -           -        -                  -   
2008    -      -        1.0      -           -        -                  -   
2009    -      -        3.0    1.0           -        -                  -   
2010    -    1.0          -      -           -        -                1.0   
2012    -      -        1.0      -           -        -                  -   
2015    -    1.0          -      -         1.0        -                  -   
2016    -    4.0          -      -           -        -                  -   
2017    -    2.0          -      -         1.0      1.0                  -   
2018    -    7.0        6.0      -           -        -                  -   

tag  computers  
year            
2006       1.0  
2008         -  
2009         -  
2010         -  
2012         -  
2015         -  
2016         -  
2017         -  
2018         -